### PR TITLE
[quant] Quantized Tensor support deepcopy

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#28612 [quant] Quantized Tensor support copy**

Summary:
att

Test Plan:
python test/test_quantized_tensor.py

Reviewers:
gchanan

Subscribers:

Tasks:

Tags:

Differential Revision: [D18255247](https://our.internmc.facebook.com/intern/diff/D18255247)

### CODE REVIEW

**Review by gchanan — COMMENTED** [review]
can you remind me why quantized tensors use qint8 storages and not int8 storages?

**Comment by jerryzh168** [comment]
> can you remind me why quantized tensors use qint8 storages and not int8 storages?

I don't remember exactly, but one reason I can think of right now is because when we inspect the storage we want to distinguish between qint8 and int8 storage?

**Review by gchanan — COMMENTED** [review]
does this support copy_, i.e. `a.copy_(b)`?

**Comment by jerryzh168** [comment]
> does this support copy_, i.e. `a.copy_(b)`?

Yes it does, test added

File: test/test_quantized_tensor.py [hunk 1]
```diff
@@ -332,5 +333,18 @@ def test_qscheme_pickle(self):
...
 
+    def test_copy(self):
+        # deep copy
+        scale, zero_point, dtype = 1.0, 2, torch.uint8
+        q_int = torch.randint(0, 100, [3, 5], dtype=dtype)
+        q = torch._make_per_tensor_quantized_tensor(q_int, scale=scale, zero_point=zero_point)
+        qc = deepcopy(q)
+        self.assertEqual(qc, q)
+        # copy_
+        qd = torch._empty_affine_quantized([3, 5], scale=scale, zero_point=zero_point, dtype=torch.quint8)
```
Comment (gchanan) [inline test/test_quantized_tensor.py]: what if the scale, zero_point, dtype, type of quantization, etc are different?

Comment (jerryzh168) [reply test/test_quantized_tensor.py]: code is here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Copy.cpp#L118-L123
it requires same quantization scheme and dtype, but allows to copy from different scale/zero_point.

Comment (jerryzh168) [reply test/test_quantized_tensor.py]: these are already tested in `test_qtensor_copy` actually, I'll update the test

**Comment by jerryzh168** [comment]
@gchanan this PR only adds support for deepcopy, normal quantized tensor copy is already supported in another PR: https://github.com/pytorch/pytorch/pull/21852

**Review by gchanan — APPROVED** [review]
this, in-and-of-itself, looks fine for deepcopy.

Why was the existing copy behavior (of clobbering the quantizer of the destination) chosen?  That seems surprising, and I don't know why you even want to copy in that case.

**Comment by jerryzh168** [comment]
> this, in-and-of-itself, looks fine for deepcopy.
> 
> Why was the existing copy behavior (of clobbering the quantizer of the destination) chosen? That seems surprising, and I don't know why you even want to copy in that case.

It's because of: https://github.com/pytorch/pytorch/blob/e5d6b753192cde81148d812ad52426e51e89e903/torch/nn/modules/module.py#L771

**Pull request closed without merging**